### ISSUE TITLE
Installation issue: make check fails

### ISSUE THREAD

**Issue by trevor-clarke** [opencog/link-grammar#1119]
Hello there!

I have spent quite some time trying to get the bindings to python working, but after googling and looking at closed issues, nothing seems to have solved my problem.

My main problem is that I can get python to find the library. Upon searching the dist-packages of my python versions, none of them contain the link grammar library despite installing the program. 

Some important things to note: 
make check, does have some issues, below are both the output from make check and from the log file it references.

Additionally, upon inspection of the log file, it mentions not having sql lite, which I installed later, make cleaned, and re-configured the program and yet the same error persists. 

TLDR:

The main issue is that the package is not getting installed to python (if that is how that works). 

Im running this on an ubuntu machine. 

I have pyton 2.7, 3.6 and 3.8 installed. 

./configure point to the 3.6 version. 

```
Making check in link-grammar
make[1]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/link-grammar'
make[2]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/link-grammar'
make[2]: Nothing to be done for 'check-am'.
make[2]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/link-grammar'
make[1]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/link-grammar'
Making check in bindings
make[1]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings'
Making check in ocaml
make[2]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/ocaml'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/ocaml'
Making check in python
make[2]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python'
Making check in python-examples
make[2]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples'
make  check-TESTS
make[3]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples'
make[4]: Entering directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples'
FAIL: tests.py
============================================================================
Testsuite summary for link-grammar 5.8.0
============================================================================
# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See bindings/python-examples/test-suite.log
Please report to https://github.com/opencog/link-grammar
============================================================================
Makefile:600: recipe for target 'test-suite.log' failed
make[4]: *** [test-suite.log] Error 1
make[4]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples'
Makefile:706: recipe for target 'check-TESTS' failed
make[3]: *** [check-TESTS] Error 2
make[3]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples'
Makefile:775: recipe for target 'check-am' failed
make[2]: *** [check-am] Error 2
make[2]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples'
Makefile:438: recipe for target 'check-recursive' failed
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings'
Makefile:550: recipe for target 'check-recursive' failed
make: *** [check-recursive] Error 1
trevor@Sketch-This-V3:~/grammar2/link-grammar/5.8.0/link-grammar-5.8.0$

```

And the output of that log file is as follows:

```
=================================================================
   link-grammar 5.8.0: bindings/python-examples/test-suite.log
=================================================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: tests
===========

Running by: /usr/bin/python3
Running ./tests.py in: /home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples
PYTHONPATH=./../python:../python:../python/.libs
srcdir=.
LINK_GRAMMAR_DATA=./../../data
Compiled with: gcc __VERSION__="7.5.0"
OS: linux-gnu __unix__
Standards: __STDC_VERSION__=201112L
Configuration (source code):
        CPPFLAGS=
        CFLAGS=-D_DEFAULT_SOURCE -std=c11 -D_BSD_SOURCE -D_SVID_SOURCE -D_GNU_SOURCE -D_ISOC11_SOURCE -g -O2 -O3
Configuration (features):
        DICTIONARY_DIR=/usr/local/share/link-grammar
        -DPACKAGE_NAME="link-grammar" -DPACKAGE_TARNAME="link-grammar" -DPACKAGE_VERSION="5.8.0" -DPACKAGE_STRING="link-grammar 5.8.0" -DPACKAGE_BUGREPORT="https://github.com/opencog/link-grammar" -DPACKAGE_URL="https://www.abisource.com/projects/link-grammar" -DPACKAGE="link-grammar" -DVERSION="5.8.0" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=".libs/" -DHAVE_STRNDUP=1 -DHAVE_STRTOK_R=1 -DHAVE_ALIGNED_ALLOC=1 -DHAVE_POSIX_MEMALIGN=1 -DHAVE_ALLOCA_H=1 -DHAVE_ALLOCA=1 -DHAVE_FORK=1 -DHAVE_VFORK=1 -DHAVE_WORKING_VFORK=1 -DHAVE_WORKING_FORK=1 -DHAVE_PRCTL=1 -D__STDC_FORMAT_MACROS=1 -D__STDC_LIMIT_MACROS=1 -DTLS=_Thread_local -DHAVE_LOCALE_T_IN_LOCALE_H=1 -DHAVE_STDATOMIC_H=1 -DUSE_WORDGRAPH_DISPLAY=1 -DHAVE_REGEX_H=1 -DHAVE_REGEXEC=1 -DHAVE_MAYBE_UNINITIALIZED=1 -DHAVE_DECL_STRERROR_R=1 -DHAVE_STRERROR_R=1 -DSTRERROR_R_CHAR_P=1
Using <module 'linkgrammar' from '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python/linkgrammar.py'>
Using <module 'clinkgrammar' from '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python/clinkgrammar.py'>
Using <module '_clinkgrammar' from '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python/.libs/_clinkgrammar.so'>
Using <module 'lg_testutils' from '/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples/lg_testutils.py'>
...............................s............s...............slink-grammar: Error: Could not open dictionary "demo-sql/4.0.dict"
E...s....s.Debug: Dictionary "he/4.0.dict": Locale "he_IL.UTF-8" unknown
link-grammar: Warning: Couldn't set dictionary locale! Using current program locale "en_US.UTF-8"
.Debug: Dictionary "ru/4.0.dict": Locale "ru_RU.UTF-8" unknown
link-grammar: Warning: Couldn't set dictionary locale! Using current program locale "en_US.UTF-8"
.ssDebug: Dictionary "de/4.0.dict": Locale "de_DE.UTF-8" unknown
link-grammar: Warning: Couldn't set dictionary locale! Using current program locale "en_US.UTF-8"
............Debug: Dictionary "lt/4.0.dict": Locale "lt_LT.UTF-8" unknown
link-grammar: Warning: Couldn't set dictionary locale! Using current program locale "en_US.UTF-8"
.Debug: Dictionary "ru/4.0.dict": Locale "ru_RU.UTF-8" unknown
link-grammar: Warning: Couldn't set dictionary locale! Using current program locale "en_US.UTF-8"
.....
======================================================================
ERROR: setUpClass (__main__.GSQLDictTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./tests.py", line 800, in setUpClass
    cls.d, cls.po = Dictionary(lang='demo-sql'), ParseOptions()
  File "/home/trevor/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python/linkgrammar.py", line 387, in __init__
    raise LG_DictionaryError('Error: Failed to open dictionary {!r}'.format(lang))
linkgrammar.LG_DictionaryError: Error: Failed to open dictionary 'demo-sql'

----------------------------------------------------------------------
Ran 91 tests in 10.386s

FAILED (errors=1, skipped=7)

trevor@Sketch-This-V3:~/grammar2/link-grammar/5.8.0/link-grammar-5.8.0/bindings/python-examples$

```

Other maybe helpful details:

![image](https://user-images.githubusercontent.com/13138669/100923696-92b11f00-34ad-11eb-87d0-489182de8edf.png)

Thank you so much! Let me know if more information is needed.

**Comment by ampli** [comment]
Hello @trevor-clarke,

> `ERROR: setUpClass (__main__.GSQLDictTestCase)`

Unless you are going to use an SQL dictionary (I guess that you are not) you can ignore this error.

> sql lite, which I installed later

Make sure you installed the **development** package (`libsqlite3-dev`).

> Upon searching the dist-packages of my python versions, none of them contain the link grammar library despite installing the program.

Look at the output of `make install` for the place it installs `linkgrammar.py`.

**Comment by trevor-clarke** [comment]
Yes. After much time, and multiple remakes, I ended up starting with a clean system with only 1 version of python. 

After going through the build steps again, I was still faced with the same error. 

This time, when I went into the python module folders, I noticed that it installed it to the wrong folder, and created a new (incorrect folder instead) 

I forget the actual foder naming and which one was which, but it was something like, the installer put it in dist-packages when it was supposed to be in site-packages.. 

Once I copied the contents of the folder that the install had created into the preexisting python folder, it worked fine.

I will update this with all the correct info and close the issue ASAP, I just wanted to have this info logged in case others are experiencing it.

**Comment by ampli** [comment]
See also https://github.com/opencog/link-grammar/issues/1094#issuecomment-619359954 .

**Comment by linas** [comment]
@trevor-clarke Did everything resolve itself in the end, or not?  As Amir states, setting `export PYTHONPATH=/usr/local/lib/python3.whatever/dist-wherever,etc` should solve the problem.

> I noticed that it installed it to the wrong folder, and created a new (incorrect folder instead)

This is a recurring and rather annoying bug that python& the various Linux distros brought onto themselves. The key problem is that there are two (or three?) distinct locations used for installing local python packages: one with the directory `site-packages` and one with `dist-packages` in the directory name. Distros which use one usually fail to look in the other. Python itself has historically changed which it uses.  There doesn't seem to be any way for link-grammar to guess the correct path, so we use the path that is compatible with Ubuntu.

**Comment by linas** [comment]
Closing; I think all of the issues here got resolved.  Please open a new issue if problems persist!
